### PR TITLE
#21446: Completion events must not be swallowed if chasing

### PR DESCRIPTION
**Pull request by drewhk**
Cancel-after-pull and Fail-after-push was encoded before by just flipping an extra bit, and when we decoded the event from the event queue we noticed the flag and treated the original push as a fail, and the original pull as cancel. With chasing, this decoding step is avoided for performance though, but this means that if we flip these bits (cancel, fail) _after_ the push or pull is registered for chasing these events are not picked up as we don't touch the queue and don't do full decoding. The events remain a pull and push, and nothing enqueues the cancel or fail.

The solution is to simply check in cancel and fail that a chasing is in progress, and revert to normal event-queue based propagation.

Complete-after-push is different and does not fail because how it is encoded (special marker element in the slot of the connection)

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 1]
```diff
@@ -842,6 +842,12 @@ final class GraphInterpreter(
       connection.portState = currentState | (OutClosed | InFailed)
       connection.slot = Failed(ex, connection.slot)
       if ((currentState & (Pulling | Pushing)) == 0) enqueue(connection)
+      else if (chasedPush eq connection) {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: the actual fix

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala [hunk 2]
```diff
@@ -853,6 +859,12 @@ final class GraphInterpreter(
     if ((currentState & OutClosed) == 0) {
       connection.slot = Empty
       if ((currentState & (Pulling | Pushing | InClosed)) == 0) enqueue(connection)
+      else if (chasedPull eq connection) {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala]: and the other fix

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterPortsSpec.scala [hunk 3]
```diff
@@ -12,1166 +12,1529 @@ class GraphInterpreterPortsSpec extends StreamSpec with GraphInterpreterSpecKit
...
+        in.isAvailable should be(false)
+        in.hasBeenPulled should be(false)
+        in.isClosed should be(true)
+        an[IllegalArgumentException] should be thrownBy {
+          in.pull()
+        }
+        an[IllegalArgumentException] should be thrownBy {
+          out.push(0)
+        }
+        an[Ille
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterPortsSpec.scala]: serious tests 👍

**Comment by johanandren** [comment]
LGTM!

**Comment by drewhk** [comment]
`?w=1` is recommended for reviewing the changes in the big red test file.

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ChasingEventsSpec.scala [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+    @throws(classOf[Exception])
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with InHandler {
+      override def preStart(): Unit = pull(in)
+      override def onPush(): Unit = pull(in)
+      setHandler(in, this)
+    }
+  }
+
+  "Even chasing" must {
```
Comment (patriknw) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ChasingEventsSpec.scala]: then I want Odd chasing ;-)
Event?

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/ChasingEventsSpec.scala]: I meant, "Even chasing must propagate cancel" ;) (thanks for noticing the typo)

**Comment by patriknw** [comment]
LGTM

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterPortsSpec.scala (hunk 3 above)
Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterPortsSpec.scala]: Yeah, once I got burnt with these corner cases I created this suite. Unfortunately it didn't end up triggering the chasing behavior. The current modifications also don't cover the things I found, I had to do more specialized test in ChasingEventsSpec, but extra coverage never hurts.

**Comment by drewhk** [comment]
merging after validation

**Comment by drewhk** [comment]
Fixes #21446
